### ISSUE TITLE
Feature Request: Local Action Hook Scripts Support

### ISSUE THREAD

**Issue by jsetton** [alexa-labs/ask-cli#41]
## I'm submitting a...

<pre><code>
[ ] Regression (a behavior that used to work and stopped working in a new release)
[ ] Bug report  
[ ] Performance issue
[x] Feature request
[ ] Documentation issue or request
[ ] Other... Please describe:
</code></pre>

## Expected Behavior
Based on my understanding, the hooks scripts are being replaced by various types of deploy delegates to provision resources on AWS. This will certainly removes the need to use the AWS CLI to provision the necessary resources via the hooks scripts. However, this doesn't seem to provide a way to run local actions prior to deploying a skill.

I have been using the hooks scripts in some of my skill development to automatically update some code configuration locally, before being zipped up, and the skill manifest as well based on a set of environment variables. This allows to quickly switch environment when deploying code directly from repository.

If this ability is indeed no longer available in the next gen ASK CLI, would it be possible to still keep that support?

**Comment by habuma** [comment]
I also enjoy using the pre-deploy hook to run automated tests before deployment. I consider ASK CLI v2 incomplete without this feature.

**Comment by RonWang** [comment]
Hey @jsetton and @habuma ,

Thanks for your great question! As you might know, we don't have any v1 hook script logic now in the skill project structure (like the pre-deploy hook) in current ask-cli-x, because we now have a **CodeBuilder** to infer what most likely is the build-flow to execute [(default built-in build-flows)](https://github.com/alexa-labs/ask-cli/tree/master/lib/builtins/build-flows). With this design, most of skill developers share similar build flows and save the effort to manage the build process.

For developers who have further desire to control the code build (you can understand it as a pre-deploy), we prepare the `custom` type of **CodeBuilder** [(details impl in here)](https://github.com/alexa-labs/ask-cli/blob/master/lib/controllers/skill-code-controller/code-builder.js#L91-L99). You just need to create a 
- Non-Windows: `{project_root}/hooks/build.sh` file
- Windows: `{project_root}\hooks\build.ps1`) file 

as your code build-flow, and ask-cli-x will execute it for all your codebase.

We will further provide document for this CodeBuilder logic before closing this issue, and for now please try it out and see if it can unblock your deployment flow.

Thanks,
Ron

**Comment by jsetton** [comment]
@RonWang thanks for your detailed explanation. I actually found out about the custom build flow, you mentioned above, a few days ago while taking a closer look at the code. From what I can see, this script is called after the skill metadata is deployed and prior to the skill infrastructure being deployed.

My initial thought was that it will handle the code configuration local updates  use case without any problem, but it wouldn't work with skill manifest changes. However, after going over the code further, I realized that the skill manifest is deployed twice; before and after the skill infrastructure is deployed. This means that any changes applied to the skill manifest during the code build phase will be deployed the second time around. I was able to confirm with my testing.

So I think the two use-cases I mentioned previously are covered by the custom build flow.

**Comment by RonWang** [comment]
You can know check the docs to understand how to manage the **Code Builder** during the deploy flow: https://github.com/alexa-labs/ask-cli/blob/master/docs/concepts/Deploy-Command.md#code-builder.

**Comment by habuma** [comment]
@RonWang Just now getting a chance to try this and have a fairly basic question...

Is it fair to say that if I create my own custom `/hooks/build.sh` script that the default `build.sh` won't be executed? And, in fact, if I create my own customer build script then I'm responsible for creating the build.zip file? 

I ask because I've created a custom script and it seems to be invoked, but `ask deploy` fails for lack of a `.ask/lambda/build.zip` file.

**Comment by jsetton** [comment]
> Is it fair to say that if I create my own custom /hooks/build.sh script that the default build.sh won't be executed? And, in fact, if I create my own customer build script then I'm responsible for creating the build.zip file?

This is correct. @RonWang pointed to the actual CodeBuilder implementation above where the `packageBuilder` path is determined. It is either uses the `custom` or `default` type.

My advice would be to build your custom script based on the default version. The same way you would use the hook scripts in the 1.x version.

**Comment by habuma** [comment]
Understood, but this differs from hooks in 1.x in that I didn't need to completely implement the zipping up of the lambda code in a 1.x pre-deploy hook, but I will need to do that here. 

I'm not saying that this is bad...just stating that it's not the same.

**Comment by jsetton** [comment]
> Understood, but this differs from hooks in 1.x in that I didn't need to completely implement the zipping up of the lambda code in a 1.x pre-deploy hook, but I will need to do that here.

I agree it's not optimal when converting from 1.x, but when you built your 1.x hooks, you started from the default template as well. At least, this is how I did it with my skill projects. So that's the analogy I was referring to.

**Comment by habuma** [comment]
Although this issue is closed, I'd offer one more thought that would make it more akin to what ASK CLI v1 had...

As it stands now, if you want custom build behavior, you must completely replace the built-in behavior with `/hooks/build.sh`, which in many cases would mean completely duplicating the existing build scripts, just to add a little something extra.

But what if there were a way to provide something like `/hooks/pre-build.sh` and/or `/hooks/post-build.sh` that, if present, would be executed before/after the built-in scripts? Then if someone wants to add some additional behavior before/after, they wouldn't need to duplicate the existing scripts.

Putting this in context, I'm totally fine with the existing build script as it is and don't wish to change it. But, before deploying anything, I want my automated tests to run. If the tests fail, I want to stop the deployment. I'm doing this just fine now with a pre-deploy hook in ASK CLI v1. But if I want to do this with ASK CLI v2, then I'd need to duplicate the existing script and inject my test run. The built-in test script is about 57 lines (excluding comments) long and I'd need to duplicate all of that to get the existing behavior and then add a handful of lines to get the testing behavior. That seems wasteful, especially since I'm otherwise completely okay with the existing behavior otherwise.

I *think* I'd likely put this new behavior in post-build.sh, so that `npm install` will have been run. But it might be useful to also support a pre-build.sh for other purposes. (I'm seriously considering an option of writing my models and skill.json in something like https://json5.org/ and the pre-processing it into the deployable JSON...not sure yet if that would fit into a pre or post build script.)

**Comment by habuma** [comment]
@RonWang Please consider the previous comment as well as this one...

Perhaps the best place to do this would be just before and just after https://github.com/alexa/ask-cli/blob/develop/lib/controllers/skill-code-controller/code-builder.js#L61-L75, as that would work regardless of which language or OS the developer is using. Alternatively, you could do it in the individual build.sh and build.ps1 scripts (in fact, I've tried this), but doing so would require duplication of the pre/post-build hooks across supported language and OS scripts. 

I do not mind working through this myself and submitting a PR. It looks fairly straightforward, albeit a bit more than I care to put into it if there's no hope of it being accepted. Let me know what you think.

**Comment by jsetton** [comment]
> My initial thought was that it will handle the code configuration local updates use case without any problem, but it wouldn't work with skill manifest changes. However, after going over the code further, I realized that the skill manifest is deployed twice; before and after the skill infrastructure is deployed. This means that any changes applied to the skill manifest during the code build phase will be deployed the second time around. I was able to confirm with my testing.

@RonWang So I should have tested my use-case of updating the skill manifest using the code builder script before posting my previous response just based on just looking at the code. Basically, this is not working because it seems that the skill manifest is cached from the initial read. Therefore any changes made by the builder script aren't taken in consideration on the second skill deployment and are actually overwritten by ask-cli when it generates the file including the Lambda endpoint URI.

First, why isn't the Lambda endpoint URI added to `.ask/ask-states.json` instead of the skill file? It makes more sense when comparing to the previous `.ask/config` file usage.

Secondly, the reason why my use-case requires building the skill file when deploying the skill is because each localized portions of the skill file are committed in separate files managed by the localization platform Crowdin. While having the ability for the ask-cli to automatically generate the skill manifest based on these files by providing an asset path to `ask-resources.json` would be a nice feature to have, I feel that revisiting the ability to have a pre-deploy hook script might be the best solution after all.

**Comment by jsetton** [comment]
@RonWang It's been a while since I posted my last comment. I was wondering what your thoughts were on the matter. Should I open a new issue?
